iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java

```diff
@@ -230,38 +229,42 @@ public int getEnd(BitReader bbv) {
 
     /**
      * The offset of the nth field depends on the length and offset of the nth-1 field. This class
-     * is used to cache static fields to avoid querying parent fields.
+     * is used to cache offset + length of static fields to avoid querying parent fields.
      *
      * Both the value and it's dynamic state are cached.
+     *
+     * Dynamic fields are not cached at the enum level and are instead resolved through the BitReader.
      */
     private static abstract class MemoizingFunction
             implements LengthSupplier, OffsetSupplier, Function<BitReader, Integer> {
-        private boolean dynamicInitialized = false;
-        private boolean isDynamic = false;
-        private Integer value;
+        private volatile boolean dynamicInitialized = false;
+        private volatile boolean isDynamic = false;
+        private volatile Integer value;
 
         public abstract Integer doCompute(BitReader t);
 
+        /**
+         * implementation must be thread-safe
+         */
         @Override
         public abstract boolean isDynamic();
 
         @Override
         public Integer apply(BitReader t) {
-            if (value == null || isDynamicPvt()) {
+            if (isDynamicPvt()) {
+                return doCompute(t);
+            } else if (value == null) {
                 value = doCompute(t);
-
-                return value;
-            } else {
-                return value;
             }
+
+            return value;
         }
 
         private boolean isDynamicPvt() {
```

Comment:
Shouldn't this method be synchronized? Or at least use something like an AtomicBoolean with a getAndSet atomic operation?

Reply:
i don't think so, what are you concerned about?

Reply:
Sorry, wrong place; I meant at line 192. I will write the comment at the proper line.